include/grass/defs/raster.h

```diff
@@ -550,8 +544,11 @@ DCELL Rast_get_d_value(const void *, RASTER_MAP_TYPE);
 /* raster_metadata.c */
 char *Rast_read_units(const char *, const char *);
 char *Rast_read_vdatum(const char *, const char *);
+char *Rast_read_bandref(const char *, const char *);
 void Rast_write_units(const char *, const char *);
 void Rast_write_vdatum(const char *, const char *);
+void Rast_write_bandref(const char *, const char *);
+int Rast_legal_bandref(const char *);
```

Comment:
Why not bool as return value type?

Reply:
Consistency.
https://github.com/OSGeo/grass/blob/019f18d61acf6e94a7afe653b13aa12a95076de6/lib/db/dbmi_base/legal_dbname.c#L31
https://github.com/OSGeo/grass/blob/019f18d61acf6e94a7afe653b13aa12a95076de6/lib/gis/legal_name.c#L34

If you want to see a real history, look into imagery lib as many functions return int just because void functions are not yet available in C at the time of writing...

Reply:
Even though I have a lot yet to learn on GRASS API, I very much aware of the situation with using int for all kind of cases, also for boolean values. Recently the possibility of using `bool` has been introduced to gis.h (see #1567 for background and related discussion). I firmly believe that code comprehension beat consistency in this case. I'm not advocating converting all boolean functions to using bool (although I'd welcome such a change), but for new code or reworked code why not use bool.


Consider:
```
int Rast_legal_bandref(const char *);
...
if (Rast_legal_bandref(bandref) < 0) {}
...
```
vs.
```
bool Rast_legal_bandref(const char *);
...
if (Rast_legal_bandref(bandref) == false) {}
...
```
the latter case is self-explanatory, both the declaration and the use of the function.

Reply:
Using bool for Rast_legal_bandref is impossible. I changed the code as you suggested, C part compiles and works just fine, but Python is a dead horse:
```
Error: dist.x86_64-pc-linux-gnu/include/grass/defs/raster.h:551: Syntax error at 'Rast_legal_bandref'
```
Unless you can spot an error in my diff, bool is a no-go:
```diff
diff --git a/include/grass/defs/raster.h b/include/grass/defs/raster.h
index 821a2fc2a..60fd4b575 100644
--- a/include/grass/defs/raster.h
+++ b/include/grass/defs/raster.h
@@ -548,7 +548,7 @@ char *Rast_read_bandref(const char *, const char *);
 void Rast_write_units(const char *, const char *);
 void Rast_write_vdatum(const char *, const char *);
 void Rast_write_bandref(const char *, const char *);
-int Rast_legal_bandref(const char *);
+bool Rast_legal_bandref(const char *);
 
 /* rast_to_img_string.c */
 int Rast_map_to_img_str(char *, int, unsigned char*);
diff --git a/lib/raster/raster_metadata.c b/lib/raster/raster_metadata.c
index c9daed88b..77e8bafa2 100644
--- a/lib/raster/raster_metadata.c
+++ b/lib/raster/raster_metadata.c
@@ -121,7 +121,7 @@ void Rast_write_bandref(const char *name, const char *str)
  * They are in format <shortcut>_<bandname>.
  * Band identifiers are capped in legth to GNAME_MAX.
  *
- * This function will return -1 if provided band id is not considered
+ * This function will return false if provided band id is not considered
  * to be valid.
  * This function does not check if band id maps to any entry in band
  * metadata files as not all band id's have files with extra metadata.
@@ -130,37 +130,37 @@ void Rast_write_bandref(const char *name, const char *str)
  *
  * \param bandref band reference to check
  *
- * \return 1 success
- * \return -1 failure
+ * \return true on success
+ * \return false on failure
  */
-int Rast_legal_bandref(const char *bandref)
+bool Rast_legal_bandref(const char *bandref)
 {
     char **tokens;
     int ntok;
 
     if (strlen(bandref) >= GNAME_MAX) {
         G_warning(_("Band reference is too long"));
-        return -1;
+        return false;
     }
 
     if (G_legal_filename(bandref) != 1)
-        return -1;
+        return false;
 
     tokens = G_tokenize(bandref, "_");
     ntok = G_number_of_tokens(tokens);
     if (ntok < 2) {
         G_warning(_("Band reference must be in form <shortcut>_<bandname>"));
         G_free_tokens(tokens);
-        return -1;
+        return false;
     }
 
     if (strlen(tokens[1]) < 1) {
         G_free_tokens(tokens);
-        return -1;
+        return false;
     }
 
     G_free_tokens(tokens);
-    return 1;
+    return true;
 }
 
 /*!
diff --git a/raster/r.support/main.c b/raster/r.support/main.c
index e1bc8472c..9bae8e5f2 100644
--- a/raster/r.support/main.c
+++ b/raster/r.support/main.c
@@ -285,7 +285,7 @@ int main(int argc, char *argv[])
     }
 
     if (bandref_opt->answer) {
-        if (Rast_legal_bandref(bandref_opt->answer) < 0)
+        if (Rast_legal_bandref(bandref_opt->answer) == false)
             G_fatal_error(_("Provided band reference is not valid. "
                             "See documentation for valid examples"));

```

Reply:
Can’t test at the moment, but I’d look into c-python bridge for possible solution.

Reply:
It seems the problem is caused by the Ctypesgen version embedded with GRASS source. It may probably be fixed, but that is for another PR. Let's keep the int version for the time being.

Reply:
Now I can confirm that the problem lies with embedded ctypesgen. I tried using (modified) generated code using  [ctypesgen](https://github.com/davidjamesca/ctypesgen) version under active development.
Inserting:
```py
for _lib in six.itervalues(_libs):
    if not hasattr(_lib, 'Rast_legal_bandref'):
        continue
    Rast_legal_bandref = _lib.Rast_legal_bandref
    Rast_legal_bandref.argtypes = [String]
    Rast_legal_bandref.restype = c_bool
    break
```
into `dist.XXX/etc/python/grass/lib/raster.py` file and the c bool type works.

I believe the long-term solution would be to update ctypesgen with upstream code, meanwhile I try to fix add bool support to existing code base.

At any rate, I withdraw this suggestion of using bool at this point.

Reply:
Reported with #1647.